lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -102,18 +108,28 @@ bool StackAnalysis<AnalyzeABI>::runOnModule(Module &M) {
     bool IsReturnAddress = hasReason(Reasons, JTReason::ReturnAddress);
     bool IsLoadAddress = hasReason(Reasons, JTReason::LoadAddress);
 
-    if (IsFunctionSymbol or IsCallee) {
-      // Called addresses are a strong hint
-      Functions.emplace_back(&BB, true);
-    } else if (not IsLoadAddress
-               and (IsUnusedGlobalData
-                    || (IsMemoryStore and not IsPCStore
-                        and not IsReturnAddress))) {
-      // TODO: keep IsReturnAddress?
-      // Consider addresses found in global data that have not been used or
-      // addresses that are not return addresses and do not end up in the PC
-      // directly.
-      Functions.emplace_back(&BB, false);
+    if (UseFunctionSymbols) {
+
+      // Rely only on debug symbols to elect CFEPs.
+      if (IsFunctionSymbol) {
+        Functions.emplace_back(&BB, true);
+      }
+    } else {
+
+      // Do the classic stuff.
+      if (IsFunctionSymbol or IsCallee) {
+        // Called addresses are a strong hint
+        Functions.emplace_back(&BB, true);
+      } else if (not IsLoadAddress
+                 and (IsUnusedGlobalData
+                      || (IsMemoryStore and not IsPCStore
+                          and not IsReturnAddress))) {
+        // TODO: keep IsReturnAddress?
+        // Consider addresses found in global data that have not been used or
+        // addresses that are not return addresses and do not end up in the PC
+        // directly.
+        Functions.emplace_back(&BB, false);
+      }
```

Comment:
What about the following?
```c++
if (IsFunctionSymbol or (not UseFunctionSymbolsOnly and IsCallee)) {
  Functions.emplace_back(&BB, true);
} else if (UseFunctionSymbolsOnly) {
  // Ignore
} else if (...) {
  // ...
  Functions.emplace_back(&BB, false);
}
```
In general, I feel like sometimes you're overly cautious in touching other code and tend to "add" features making sure you don't break other stuff, which is good in principle. However, it creates a layers and there's only one problem that cannot be solved by adding another layer, having too many layers.

Feel free to move things around, if tests pass your change is good!